### ISSUE TITLE
Using Intermediate (non-Input) Layer as input to secondary network

### ISSUE THREAD

**Issue by phillips96** [tensorflow/tensorflow#43025]
**System information**
- TensorFlow version: 2.3.0 
- Are you willing to contribute it: No

**Describe the feature and the current behavior/state.**
Currently, a user can define a network to only run partially through a network using something like below:
```
# Load Pre-Trained Model
model = ...
# Using I/O of previous model, build model with an additional (or replacement) output 
first_half = tf.keras.models.Model(model.inputs, model.get_layer("some_intermediate_layer_n").output)
```
However, if you were to build a final model to run through the remainder of the network, from `some_intermediate_layer_n` onward, it falls over. It would be nice to be able to make a similar call, to use the base models outputs, but starting from an intermediate input step as below. You could use the output from the network above as the input for the remainder of the network in this example. 
```
 second_half = tf.keras.models.Model(model.get_layer("some_intermediate_layer_n_plus_1").input, model.outputs)
```

 The exact warning is `WARNING:tensorflow:Functional inputs must come from tf.keras.Input (thus holding past layer metadata), they cannot be the output of a previous non-Input layer.` This is quickly followed by a `Graph Disconnected` error.

**Will this change the current api? How?**
Not to my knowledge. Should be able to call in a similar fashion above, where the new input is the intermediate layer you wish to start from.

**Who will benefit with this feature?**
Any network architecture where an intermediate representation is important. Architecture's reliant on pairs of networks, or encoding networks would benefit at the inference stage.  

**Any Other info.**
I'm assuming it's quite likely this has been requested before but I couldn't find it. If so, it would also be nice to have an explanation as to why the current limitation to inputs coming from `tf.keras.Input (thus holding past layer metadata)` is important. I understand this from a training perspective, but when inferring, I would've thought this could be overlooked.

I of course understand you could build two networks with the latter building on from the prior but this way you can wrap it all up into one when training, or if using a pre-trained network, you can split it out as and when needed. 

Thank you.

**Comment by bhack** [comment]
Do you have a very, very minima but runnable example or colab to reproduce this?

**Comment by phillips96** [comment]
I'll pull something together with some sample weights tomorrow! 
Thanks

**Comment by phillips96** [comment]
Please see the primitive example below;
```
import tensorflow as tf
def def_enc_dec_pair():
    input_ = tf.keras.layers.Input(shape=[None, None, 3])
    x_e = tf.keras.layers.Conv2D(filters=8, kernel_size=(3, 3), strides=(2, 2), padding="SAME")(input_)
    x_e = tf.keras.layers.Conv2D(filters=16, kernel_size=(3, 3), strides=(2, 2), padding="SAME")(x_e)
    x_e = tf.keras.layers.Conv2D(filters=8, kernel_size=(3, 3), strides=(2, 2), padding="SAME")(x_e)
    x_e = tf.keras.layers.Conv2D(filters=3, kernel_size=(3, 3), strides=(2, 2), padding="SAME", name="encoded_layer")(x_e)

    x_d = tf.keras.layers.Conv2DTranspose(filters=8, kernel_size=(3, 3), strides=(2, 2), padding="SAME", name="second_section")(x_e)
    x_d = tf.keras.layers.Conv2DTranspose(filters=16, kernel_size=(3, 3), strides=(2, 2), padding="SAME")(x_d)
    x_d = tf.keras.layers.Conv2DTranspose(filters=8, kernel_size=(3, 3), strides=(2, 2), padding="SAME")(x_d)
    x_d = tf.keras.layers.Conv2DTranspose(filters=3, kernel_size=(3, 3), strides=(2, 2), padding="SAME")(x_d)

    model = tf.keras.Model(inputs=[input_], outputs=[x_d])
    model.compile(optimizer=tf.optimizers.Adam(), loss=[tf.keras.losses.MSE])

    return model

base_model = def_enc_dec_pair()
# base_model.summary()

first_half = tf.keras.Model(base_model.inputs, base_model.get_layer("encoded_layer").output)
# first_half.summary()

second_half = tf.keras.Model(base_model.get_layer("second_section").input, base_model.output)
second_half.summary()
```

Note that for usability, it would be beneficial to call the input to `second_half` as one of the following options in this example
`base_model.get_layer("encoded_layer").output`
`base_model.get_layer("second_section").input`
`first_half.output`

I hope this clarifies exactly what i'm talking about. `base_model` and `first_half` build fine but `second_half` fails for the previously mentioned reason.

**Comment by bhack** [comment]
So it seems to me that your 2nd use case it is like: https://stackoverflow.com/questions/52800025/keras-give-input-to-intermediate-layer-and-get-final-output

**Comment by phillips96** [comment]
That looks like what i'm after yes. Using against this trivial example you'd build as such

```
...
second_input_shape = base_model.get_layer("encoded_layer").output.shape
second_input_shape = second_input_shape[1:]  # Remove Batch from front.
second_input = tf.keras.layers.Input(shape=second_input_shape)
x = second_input
for layer in base_model.layers[5:]:  # encoded_layer idx + 1
    x = layer(x)

second_half = tf.keras.Model(second_input, x)
second_half.summary()
```
For this simplistic approach, the above is okay but it could quickly get out of hand as the original stackoverflow answer points out. In my actual code i'm having to do something slightly more convoluted to stitch the remaining layers together. 

Considering the initial hurdle is simply to create an Input Layer with dimensions to match the layer we want, is it something that could still be of added benefit? 

My suggestion would be instead of throwing the warning, an attempt could be made to simply create an `Input` from the requested starting point (whether that's the output shape of a layer or the input shape of the next as they'll match up ofc), and then look at all the connections from the base network that are required to go from this point in the original model to the requested output.

**Comment by bhack** [comment]
So do you mean that you want to automate this?

**Comment by phillips96** [comment]
Basically yes

**Comment by bhack** [comment]
I don't know the recent internals design limit to automate this in the Keras API
/cc @tanzhenyu do you have an opinion about this?

**Comment by phillips96** [comment]
Just to state for the record the above solution becomes a nightmare when you're using anything residual or with multiple paths, which is why is why it would be so useful to base it off of the knowledge you already have about all the connections from the original `base_model`

**Comment by tanzhenyu** [comment]
This is where subclassed model works better than functional models:
```
class Encoder(tf.keras.Model):
  def __init__(self):
    super(Encoder, self).__init__()
    self.encoder_layers = []
    self.encoder_layers.append(tf.keras.layers.Conv2D(filters=8, kernel_size=(3, 3), strides=(2, 2), padding="SAME"))
    self.encoder_layers.append(tf.keras.layers.Conv2D(filters=16, kernel_size=(3, 3), strides=(2, 2), padding="SAME"))
    self.encoder_layers.append(tf.keras.layers.Conv2D(filters=8, kernel_size=(3, 3), strides=(2, 2), padding="SAME"))
    self.encoder_layers.append(tf.keras.layers.Conv2D(filters=3, kernel_size=(3, 3), strides=(2, 2), padding="SAME", name="encoded_layer"))

  def call(self, inputs):
    output = inputs
    for layer in self.encoder_layers:
      output = layer(output)
    return output

class Decoder(tf.keras.Model):
  def __init__(self):
    super(Decoder, self).__init__()
    self.decoder_layers = []
    self.decoder_layers.append(tf.keras.layers.Conv2DTranspose(filters=8, kernel_size=(3, 3), strides=(2, 2), padding="SAME", name="second_section"))
    self.decoder_layers.append(tf.keras.layers.Conv2DTranspose(filters=16, kernel_size=(3, 3), strides=(2, 2), padding="SAME"))
    self.decoder_layers.append(tf.keras.layers.Conv2DTranspose(filters=8, kernel_size=(3, 3), strides=(2, 2), padding="SAME"))
    self.decoder_layers.append(tf.keras.layers.Conv2DTranspose(filters=3, kernel_size=(3, 3), strides=(2, 2), padding="SAME"))
  
  def call(self, inputs):
    output = inputs
    for layer in self.decoder_layers:
      output = layer(output)
    return output


class MyModel(tf.keras.Model):
  def __init__(self):
    super(MyModel, self).__init__()
    self.encoder = Encoder()
    self.decoder = Decoder()

  def call(self, inputs):
    return self.decoder(self.encoder(inputs))
```

Making things modular will also help you even in residual path case, no?

**Comment by bhack** [comment]
@phillips96 Of course If you don't need to be constrained inside the functional API perimeter as @tanzhenyu's suggested you can handle this with model subclass.

**Comment by phillips96** [comment]
Thanks @tanzhenyu, so my actual application uses subclassed models in a similar fashion to this.

I just found it curious that the api is capable of building a sub-model going from the input to and intermediate layer using that api call, but you can't go from the intermediate layer to the final output it quite the same fashion.

Happy to close this if it's not something that seems overall beneficial to the repository 😄

**Comment by tanzhenyu** [comment]
> Thanks @tanzhenyu, so my actual application uses subclassed models in a similar fashion to this.
> 
> I just found it curious that the api is capable of building a sub-model going from the input to an intermediate layer using that api call, but you can't go from the intermediate layer to the final output in quite the same fashion.
> 
> Happy to close this if it's not something that seems overall beneficial to the repository 😄

No, I agree with you, this seems desirable, however currently blocked by the same reason mentioned in the stackflow thread. Contribution welcome!

**Comment by bhack** [comment]
@tanzhenyu Yes this was the real question. Glad to know that there is any hard design constrain to implement this.

@phillips96 I suppose that a PR is welcome if you want to implement this behaviour.

**Comment by phillips96** [comment]
@bhack sure thing.

This will be my first PR here so I'll sign up to the contributors forms and do some initial scoping for this.

**Comment by gowthamkpr** [comment]
@phillips96 Glad you were able to solve your issue. If you wanna contribute, please post this issue in [keras-team/keras](https://github.com/keras-team/keras/issues/new/choose) repo as the development of keras has been moved to a different repo. Thanks!!
